### ISSUE TITLE
Implement Rust wrappers around Value

### ISSUE THREAD

**Issue by lopopolo** [lopopolo/ferrocarril#17]
This is a tracking ticket for implementing structs that wrap a `Value` with a single `Ruby` type, e.g. a struct that wraps `mrb_vtype::MRB_TT_HASH`, `mrb_vtype::MRB_TT_ARRAY`, or `mrb_vtype::MRB_TT_DATA`.

In order to implement `Clone` safely (see GH-3), `Value`s with `Ruby::Data` Ruby type need to know their underlying `T`, but `Value` can't be generic on `T`. Wrapping `Value` and calling those wrappers from consumers of the `mruby` crate enables application code to check their invariants about the types of objects they are getting back from the mruby VM.

At a minimum, each wrapper should implement:

- `new(Value) -> Result<Self,MrbError>`
- `Clone::clone`
- `funcall(self, method: String, args: &[Value]) -> Result<Value, MrbError>` (we might be able to move this to a `ValueLike` trait and provide a default impl).
- `funcall_with_block(self, method: String, args: &[Value], blk: Proc) -> Result<Value, MrbError>` (out of scope).

These wrappers are also a convenient place to implement the methods exposed by the mruby VM on Rust types, e.g. `String#sub`.

This might be a means to implement a `Proc` wrapper in terms of Rust closures, but that is out of scope for this tracking ticket.

Child tickets:

- `Ruby::Data`: GH-64
- `Ruby::Array`: GH-65

**Comment by lopopolo** [comment]
blocked on GH-46

**Comment by lopopolo** [comment]
This ticket has been repurposed from being about implementing wrappers around `mrb_funcall_argv` to encompass wrapping all `mrb_vtype`s with one Rust struct per `Ruby` type to accommodate implementing `Clone`, see GH-3.
